psd-web/app/models/user.rb

```diff
@@ -181,9 +185,54 @@ def invitation_expired?
     invited_at <= INVITATION_EXPIRATION_DAYS.days.ago
   end
 
-private
+  def two_factor_authentication_code_expired?
+    return false if !direct_otp_sent_at
+
+    (direct_otp_sent_at + User.direct_otp_valid_for) < Time.current
+  end
+
+  def two_factor_locked?
+    second_factor_attempts_locked_at && !two_factor_lock_expired?
+  end
+
+  def fail_two_factor_authentication!
+    return if max_login_attempts?
+
+    self.second_factor_attempts_count += 1
+    self.save
+    lock_two_factor! if max_login_attempts?
+  end
+
+  def pass_two_factor_authentication!
+    unlock_two_factor! if max_login_attempts?
+    update_column(:second_factor_attempts_count, 0)
+  end
 
   # BEGIN: place devise overriden method calls bellow
+  def send_two_factor_authentication_code(code)
+    SendTwoFactorAuthenticationJob.perform_later(self, code)
+  end
+
+  def need_two_factor_authentication?(_request)
+    Rails.configuration.two_factor_authentication_enabled
+  end
+
+private
+
+  def lock_two_factor!
+    self.update_column(:second_factor_attempts_locked_at, Time.current)
+  end
+
+  def unlock_two_factor!
+    self.update_column(:second_factor_attempts_locked_at, nil)
+  end
+
+  def two_factor_lock_expired?
+    return true if !second_factor_attempts_locked_at
```

Comment:
Suggestion for readability:

```suggestion
    return true if second_factor_attempts_locked_at.nil?
```